src/BSON/Value.c

```diff
@@ -0,0 +1,631 @@
+}
+
+static PHP_METHOD(MongoDB_BSON_Value, getSymbol)
+{
+	php_phongo_value_t* intern;
+
+	PHONGO_PARSE_PARAMETERS_NONE();
+
+	intern = Z_VALUE_OBJ_P(getThis());
+
+	PHONGO_VALUE_CHECK_TYPE(intern->value, BSON_TYPE_SYMBOL);
+
+	php_phongo_bson_new_symbol(return_value, intern->value.value.v_symbol.symbol, intern->value.value.v_symbol.len);
+}
+
+static PHP_METHOD(MongoDB_BSON_Value, getTimestamp)
+{
+	php_phongo_value_t* intern;
+
+	PHONGO_PARSE_PARAMETERS_NONE();
+
+	intern = Z_VALUE_OBJ_P(getThis());
+
+	PHONGO_VALUE_CHECK_TYPE(intern->value, BSON_TYPE_TIMESTAMP);
+
+	php_phongo_bson_new_timestamp(return_value, intern->value.value.v_timestamp.increment, intern->value.value.v_timestamp.timestamp);
+}
+
+static PHP_METHOD(MongoDB_BSON_Value, getUndefined)
+{
+	php_phongo_value_t* intern;
+
+	PHONGO_PARSE_PARAMETERS_NONE();
+
+	intern = Z_VALUE_OBJ_P(getThis());
+
+	PHONGO_VALUE_CHECK_TYPE(intern->value, BSON_TYPE_UNDEFINED);
+
+	object_init_ex(return_value, php_phongo_undefined_ce);
+}
+
+static PHP_METHOD(MongoDB_BSON_Value, getUTCDateTime)
+{
+	php_phongo_value_t* intern;
+
+	PHONGO_PARSE_PARAMETERS_NONE();
+
+	intern = Z_VALUE_OBJ_P(getThis());
+
+	PHONGO_VALUE_CHECK_TYPE(intern->value, BSON_TYPE_DATE_TIME);
+
+	php_phongo_bson_new_utcdatetime(return_value, intern->value.value.v_datetime);
+}
+
+static PHP_METHOD(MongoDB_BSON_Value, getUtf8)
+{
+	php_phongo_value_t* intern;
+
+	PHONGO_PARSE_PARAMETERS_NONE();
+
+	intern = Z_VALUE_OBJ_P(getThis());
+
+	PHONGO_VALUE_CHECK_TYPE(intern->value, BSON_TYPE_UTF8);
+
+	RETURN_STRINGL(intern->value.value.v_utf8.str, intern->value.value.v_utf8.len);
+}
+
+PHONGO_VALUE_TYPE_ISSER(isArray, BSON_TYPE_ARRAY);
+PHONGO_VALUE_TYPE_ISSER(isBinary, BSON_TYPE_BINARY);
+PHONGO_VALUE_TYPE_ISSER(isBool, BSON_TYPE_BOOL);
+PHONGO_VALUE_TYPE_ISSER(isDBPointer, BSON_TYPE_DBPOINTER);
+PHONGO_VALUE_TYPE_ISSER(isDecimal128, BSON_TYPE_DECIMAL128);
+PHONGO_VALUE_TYPE_ISSER(isDouble, BSON_TYPE_DOUBLE);
+PHONGO_VALUE_TYPE_ISSER(isDocument, BSON_TYPE_DOCUMENT);
+PHONGO_VALUE_TYPE_ISSER(isInt32, BSON_TYPE_INT32);
+PHONGO_VALUE_TYPE_ISSER(isInt64, BSON_TYPE_INT64);
+PHONGO_VALUE_TYPE_ISSER(isMaxKey, BSON_TYPE_MAXKEY);
+PHONGO_VALUE_TYPE_ISSER(isMinKey, BSON_TYPE_MINKEY);
+PHONGO_VALUE_TYPE_ISSER(isNull, BSON_TYPE_NULL);
+PHONGO_VALUE_TYPE_ISSER(isObjectId, BSON_TYPE_OID);
+PHONGO_VALUE_TYPE_ISSER(isRegex, BSON_TYPE_REGEX);
+PHONGO_VALUE_TYPE_ISSER(isSymbol, BSON_TYPE_SYMBOL);
+PHONGO_VALUE_TYPE_ISSER(isTimestamp, BSON_TYPE_TIMESTAMP);
+PHONGO_VALUE_TYPE_ISSER(isUndefined, BSON_TYPE_UNDEFINED);
+PHONGO_VALUE_TYPE_ISSER(isUTCDateTime, BSON_TYPE_DATE_TIME);
+PHONGO_VALUE_TYPE_ISSER(isUtf8, BSON_TYPE_UTF8);
+
+static PHP_METHOD(MongoDB_BSON_Value, isCode)
+{
+	php_phongo_value_t* intern;
+
+	PHONGO_PARSE_PARAMETERS_NONE();
+
+	intern = Z_VALUE_OBJ_P(getThis());
+
+	RETURN_BOOL(intern->value.value_type == BSON_TYPE_CODE || intern->value.value_type == BSON_TYPE_CODEWSCOPE);
```

Comment:
This is the only case (apart from the `int` and `number` convenience helpers) where we return the same class instance for two different BSON types, hence the explicit declaration instead of using `PHONGO_VALUE_TYPE_ISSER`.

Reply:
I believe your comment refers to encoding a `Javascript` instance as BSON? I didn't touch that part yet (instead relying on `zval_to_bson` to encode objects), but I agree that we should ensure to not use `CODEWSCOPE` if the `Javascript` instance does not contain a scope.

Reply:
I think the impetus for that last comment was realizing that `getCode()` was essentially an alias like `getInt()`. I was considering how users might want to determine the original BSON type given just the returned Javascript object. AFAIK, the only way to do so is checking whether `getScope()` is `null`.

I don't think there's anything to do here other than make sure we document how `getCode()` works by allowing both types.